pkg/gcp/actuator/actuator.go

```diff
@@ -0,0 +1,597 @@
+	annotatorconst "github.com/openshift/cloud-credential-operator/pkg/controller/secretannotator/constants"
+	"github.com/openshift/cloud-credential-operator/pkg/controller/utils"
+	ccgcp "github.com/openshift/cloud-credential-operator/pkg/gcp"
+
+	// GCP packages
+	iamadminpb "google.golang.org/genproto/googleapis/iam/admin/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	rootGCPCredsSecretNamespace = "kube-system"
+	rootGCPCredsSecret          = "gcp-creds"
+	roGCPCredsSecretNamespace   = "openshift-cloud-credential-operator"
+	roGCPCredsSecret            = "cloud-credential-operator-iam-ro-creds"
+
+	gcpSecretJSONKey = "serviceaccount.json"
+)
+
+var _ actuatoriface.Actuator = (*Actuator)(nil)
+
+// Actuator implements the CredentialsRequest Actuator interface to create credentials for GCP.
+type Actuator struct {
+	Client           client.Client
+	Codec            *minterv1.ProviderCodec
+	GCPClientBuilder func([]byte) (ccgcp.Client, error)
+}
+
+// NewActuator initializes and returns a new Actuator for GCP.
+func NewActuator(c client.Client) (*Actuator, error) {
+	codec, err := minterv1.NewCodec()
+	if err != nil {
+		log.WithError(err).Error("error creating GCP codec")
+		return nil, fmt.Errorf("error creating GCP codec: %v", err)
+	}
+
+	return &Actuator{
+		Client:           c,
+		Codec:            codec,
+		GCPClientBuilder: ccgcp.NewClient,
+	}, nil
+}
+
+// Create the credentials.
+func (a *Actuator) Create(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	return a.sync(ctx, cr)
+}
+
+// Delete the credentials. If no error returned, it is assumed that all dependent resources have been cleaned up.
+func (a *Actuator) Delete(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	if isGCP, err := isGCPCredentials(cr.Spec.ProviderSpec); !isGCP {
+		return err
+	}
+	logger := a.getLogger(cr)
+	logger.Debug("running Delete")
+
+	gcpStatus, err := decodeProviderStatus(a.Codec, cr)
+	if err != nil {
+		return err
+	}
+
+	if gcpStatus.ServiceAccountID == "" {
+		logger.Warn("no service account ID set on credentialsRequest, provisioned via passthrough")
+		return nil
+	}
+	logger = logger.WithField("serviceAccountID", gcpStatus.ServiceAccountID)
+
+	logger.Info("deleting service account from GCP")
+
+	gcpClient, err := a.buildRootGCPClient(cr)
+	if err != nil {
+		return err
+	}
+
+	svcAcct, err := getServiceAccount(gcpClient, gcpStatus.ServiceAccountID)
+	if err != nil {
+		return fmt.Errorf("error getting service account details: %v", err)
+	}
+
+	if err := removeAllPolicyBindingsFromServiceAccount(gcpClient, svcAcct); err != nil {
+		return fmt.Errorf("error removing service account policy bindings: %v", err)
+	}
```

Comment:
Is this required before it will let you delete the service account?

Reply:
Sadly, yes. GCP lets you remove a service account without undoing the policy bindings. For example, in the following output the serviceaccount "test-serviceaccount" has been deleted, but the policy binding still exists.
```
[jdiaz@minigoomba ~]$ gcloud projects get-iam-policy openshift-hive-dev --format=json | jq  '.bindings[] | select(.role == "roles/appengine.codeViewer")'
{
  "members": [
    "serviceAccount:test-serviceaccount@openshift-hive-dev.iam.gserviceaccount.com"
  ],
  "role": "roles/appengine.codeViewer"
}
```
So this policy cleanup is needed to avoid leaving cruft around GCP.

Reply:
I guess I didn't really answer your question. It is **not** required to remove the role bindings from the policy before deleting the service account, but it will then leave the role binding entries for the now non-existent service account.
So we really should be cleaning up the policy before removing the service account.